File: frontend/src/utils/permissions.ts
/**
 * Frontend permission helpers mirroring backend/app/core/permissions.py
 * for UI gating. Server enforcement remains authoritative.
 */
import type { AppRole } from '@store/authStore';

type Action =
  | 'view'
  | 'create'
  | 'edit'
  | 'delete'
  | 'upload'
  | 'approve'
  | 'reject'
  | 'publish'
  | 'assign';

const ADMIN_ACTIONS: Action[] = [
  'view', 'create', 'edit', 'delete', 'upload', 'approve', 'reject', 'publish', 'assign',
];

const MATRIX: Record<string, Partial<Record<AppRole, Action[]>>> = {
  users: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    // Managers manage field engineers via dedicated endpoints (not this matrix alone).
    manager: ['view', 'edit', 'assign'],
  },
  projects: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    manager: ['view'],
    field_engineer: ['view'],
  },
  towers: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    manager: ['view'],
    field_engineer: ['view'],
  },
  floors: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    manager: ['view'],
    field_engineer: ['view'],
  },
  flats: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    manager: ['view'],
    field_engineer: ['view'],
  },
  rooms: {
    admin: ADMIN_ACTIONS,
    super_admin: ADMIN_ACTIONS,
    manager: ['view'],
    field_engineer: ['view', 'create', 'edit', 'delete'],
  },
  captures: {
    admin: ['view', 'create', 'edit', 'delete', 'approve', 'reject', 'upload'],
    super_admin: ['view', 'create', 'edit', 'delete', 'approve', 'reject', 'upload'],
    manager: ['view', 'approve', 'reject'],
    field_engineer: ['view', 'upload', 'create', 'edit', 'delete'],
  },
  tours: {
    admin: ['view', 'create', 'edit', 'delete', 'publish'],
    super_admin: ['view', 'create', 'edit', 'delete', 'publish'],
    manager: ['view', 'publish'],
    field_engineer: ['view', 'create', 'publish', 'edit', 'delete'],
  },
  floorPlans: {
    admin: ['view', 'create', 'edit', 'delete', 'upload'],
    super_admin: ['view', 'create', 'edit', 'delete', 'upload'],
    manager: ['view', 'create', 'edit', 'delete', 'upload'],
    field_engineer: ['view', 'create', 'upload', 'edit'],
  },
  defects: {
    admin: ['view', 'create', 'edit', 'delete'],
    super_admin: ['view', 'create', 'edit', 'delete'],
    manager: ['view', 'create', 'edit'],
    field_engineer: ['view'],
  },
  analytics: {
    admin: ['view'],
    super_admin: ['view'],
    manager: ['view'],
  },
  settings: {
    admin: ['view', 'edit'],
    super_admin: ['view', 'edit'],
  },
  organizations: {
    admin: ['view', 'edit'],
    super_admin: ['view', 'edit'],
  },
  auditLogs: {
    admin: ['view'],
    super_admin: ['view'],
  },
  media: {
    admin: ['view'],
    super_admin: ['view'],
  },
  notifications: {
    admin: ['view', 'create', 'edit', 'delete'],
    super_admin: ['view', 'create', 'edit', 'delete'],
    manager: ['view', 'create', 'edit', 'delete'],
    field_engineer: ['view', 'create', 'edit', 'delete'],
  },
};

export function can(
  role: AppRole | undefined,
  module: keyof typeof MATRIX,
  action: Action,
): boolean {
  if (!role) return false;
  const allowed = MATRIX[module]?.[role];
  return !!allowed?.includes(action);
}

File: backend/app/core/permissions.py
"""
System-role permission matrix — the SERVER-SIDE source of truth for what each
system role may do on each module.

This mirrors the frontend matrix in `frontend/src/utils/permissions.ts` exactly.
The frontend gates the UI; this module enforces the same rules on the API so a
crafted request (or a stolen low-privilege token) cannot perform an action the
UI would never offer. Keep the two matrices in sync.

Scope note: this is SYSTEM-role enforcement (users.role), the outer gate. The
finer project-scoped RBAC (user_projects) remains available via RBACService for
endpoints that carry a project_id; this matrix closes the broad hole where every
authenticated user could write/delete any entity in their org regardless of role.
"""
from typing import Optional

from fastapi import Depends

from app.core.dependencies import OrgContext, get_org_context
from app.core.exceptions import ForbiddenException

# ── Action / module vocabulary (mirrors the frontend) ─────────────────────────
Action = str   # 'view'|'create'|'edit'|'delete'|'approve'|'reject'|'publish'|'upload'|'assign'
Module = str

_PERMISSIONS: dict[str, dict[Module, set[Action]]] = {
    "super_admin": {
        "users":         {"view", "create", "edit", "delete", "assign"},
        "projects":      {"view", "create", "edit", "delete", "assign"},
        "towers":        {"view", "create", "edit", "delete"},
        "floors":        {"view", "create", "edit", "delete"},
        "flats":         {"view", "create", "edit", "delete"},
        "rooms":         {"view", "create", "edit", "delete"},
        "captures":      {"view", "create", "edit", "delete", "approve", "reject", "upload"},
        "tours":         {"view", "create", "edit", "delete", "publish"},
        "floorPlans":    {"view", "create", "edit", "delete", "upload"},
        "defects":       {"view", "create", "edit", "delete"},
        "analytics":     {"view"},
        "settings":      {"view", "edit"},
        "organizations": {"view", "edit"},
        "auditLogs":     {"view"},
        "notifications": {"view", "create", "edit", "delete"},
    },
    "admin": {
        "users":         {"view", "create", "edit", "delete", "assign"},
        "projects":      {"view", "create", "edit", "delete", "assign"},
        "towers":        {"view", "create", "edit", "delete"},
        "floors":        {"view", "create", "edit", "delete"},
        "flats":         {"view", "create", "edit", "delete"},
        "rooms":         {"view", "create", "edit", "delete"},
        "captures":      {"view", "create", "edit", "delete", "approve", "reject", "upload"},
        "tours":         {"view", "create", "edit", "delete", "publish"},
        "floorPlans":    {"view", "create", "edit", "delete", "upload"},
        "defects":       {"view", "create", "edit", "delete"},
        "analytics":     {"view"},
        "settings":      {"view", "edit"},
        "organizations": {"view", "edit"},
        "auditLogs":     {"view"},
        "notifications": {"view", "create", "edit", "delete"},
    },
    "manager": {
        "projects":      {"view"},
        "towers":        {"view"},
        "floors":        {"view"},
        "flats":         {"view"},
        "rooms":         {"view"},
        "captures":      {"view", "approve", "reject"},
        "tours":         {"view", "publish"},
        "floorPlans":    {"view"},
        "defects":       {"view", "create", "edit"},
        "analytics":     {"view"},
        "notifications": {"view", "create", "edit", "delete"},
    },
    # 'reviewer' is treated like a manager for capture review in the product.
    "reviewer": {
        "projects":      {"view"},
        "towers":        {"view"},
        "floors":        {"view"},
        "flats":         {"view"},
        "rooms":         {"view"},
        "captures":      {"view", "approve", "reject"},
        "tours":         {"view"},
        "floorPlans":    {"view"},
        "defects":       {"view", "create", "edit"},
        "analytics":     {"view"},
        "notifications": {"view", "create", "edit", "delete"},
    },
    "field_engineer": {
        "projects":      {"view"},
        "towers":        {"view"},
        "floors":        {"view"},
        "flats":         {"view"},
        # pins/rooms are created AND renamed/resequenced as part of the capture
        # + pin-delete workflow, so field engineers need create+edit+delete here.
        "rooms":         {"view", "create", "edit", "delete"},
        "captures":      {"view", "upload", "create", "edit", "delete"},
        "floorPlans":    {"view", "create", "upload", "edit"},
        # Field engineers publish floor-plan walkthroughs from /publish-tours
        # (a FieldEngineerRoute) and can delete a walkthrough from /tours (the
        # delete control there is not role-gated), so they need full tour CRUD.
        # NOTE: this intentionally diverges from the stale frontend matrix, which
        # lists tours:[] for field_engineer but is contradicted by the actual UI.
        "tours":         {"view", "create", "publish", "edit", "delete"},
        "defects":       {"view"},
        "notifications": {"view", "create", "edit", "delete"},
    },
    # Generic low-privilege authenticated user / viewer: read-only.
    "user": {
        "projects": {"view"}, "towers": {"view"}, "floors": {"view"}, "flats": {"view"},
        "rooms": {"view"}, "captures": {"view"}, "tours": {"view"}, "floorPlans": {"view"},
        "defects": {"view"}, "analytics": {"view"}, "notifications": {"view", "edit", "delete"},
    },
    "viewer": {
        "projects": {"view"}, "towers": {"view"}, "floors": {"view"}, "flats": {"view"},
        "rooms": {"view"}, "captures": {"view"}, "tours": {"view"}, "floorPlans": {"view"},
        "defects": {"view"}, "analytics": {"view"}, "notifications": {"view"},
    },
}


def can(role: Optional[str], module: Module, action: Action) -> bool:
    """Returns True if the system role permits the action on the module."""
    if not role:
        return False
    return action in _PERMISSIONS.get(role, {}).get(module, set())


def require_permission(module: Module, action: Action):
    """
    FastAPI dependency factory enforcing a system-role permission.

    Usage:
        @router.delete("/projects/{project_id}")
        async def delete_project(..., _=Depends(require_permission("projects", "delete"))):
    """
    async def _check(ctx: OrgContext = Depends(get_org_context)) -> OrgContext:
        if not can(ctx.role, module, action):
            raise ForbiddenException(
                f"Your role ('{ctx.role}') is not permitted to {action} {module}."
            )
        return ctx

    return _check
